semmc/tests/semstore/ParamFormulaTests.hs

```diff
@@ -0,0 +1,316 @@
+import           Lang.Crucible.Backend.Simple ( newSimpleBackend )
+import qualified SemMC.BoundVar as BV
+import qualified SemMC.Formula.Formula as SF
+import qualified SemMC.Formula.Parser as FI
+import qualified SemMC.Formula.Printer as FO
+import qualified SemMC.Log as Log
+import           Test.Tasty
+import           Test.Tasty.Hedgehog
+import           TestArch
+import           TestArchPropGen
+import           TestUtils
+import           What4.BaseTypes
+
+import           Prelude
+
+
+parameterizedFormulaTests :: [TestTree]
+parameterizedFormulaTests = [
+  testGroup "Parameterized Formulas" $
+
+    [ testProperty "parameter type" $
+      property $ do Some r <- liftIO newIONonceGenerator
+                    sym <- liftIO $ newSimpleBackend r
+                    (p, _operands) <- forAllT (genParameterizedFormula sym OpSurf)
+                    assert (all isValidParamType (SF.pfUses p))
+    , testProperty "parameter type multiple" $
+      property $ do Some r <- liftIO newIONonceGenerator
+                    sym <- liftIO $ newSimpleBackend r
+                    (p, _operands) <- forAllT (genParameterizedFormula sym OpPack)
+                    assert (all isValidParamType (SF.pfUses p))
+    , testProperty "operand type" $
+      property $ do Some r <- liftIO newIONonceGenerator
+                    sym <- liftIO $ newSimpleBackend r
+                    (p, _operands) <- forAllT (genParameterizedFormula sym OpSurf)
+                    assert $ isNatArgFoo ((SF.pfOperandVars p) SL.!! SL.index0)
+    , testProperty "literal vars" $
+      property $ do Some r <- liftIO newIONonceGenerator
+                    sym <- liftIO $ newSimpleBackend r
+                    _ <- forAllT (genParameterizedFormula sym OpSurf)
+                    success -- TBD: something (manything?) to test literal vars here
+      -- TBD: needs other tests
+    , testProperty "defs keys in uses" $
+      property $ do Some r <- liftIO newIONonceGenerator
+                    sym <- liftIO $ newSimpleBackend r
+                    (p, _operands) <- forAllT (genParameterizedFormula sym OpSurf)
+                    assert (all (flip Set.member (SF.pfUses p)) (MapF.keys $ SF.pfDefs p))
+
+    , testProperty "serialized formula round trip, simple backend, OpPack" $
+      property $ do Some r <- liftIO newIONonceGenerator
+                    sym <- liftIO $ newSimpleBackend r
+                    let opcode = OpPack
+                    (p, _operands) <- forAllT (genParameterizedFormula sym opcode)
+                    debugPrint $ "parameterizedFormula: " <> show p
+                    debugPrint $ "# literalVars: " <> show (MapF.size $ SF.pfLiteralVars p)
+                    debugPrint $ "# defs: " <> show (MapF.size $ SF.pfDefs p)
+                    let printedFormula = FO.printParameterizedFormula (HR.typeRepr opcode) p
+                    debugPrint $ "printedFormula: " <> show printedFormula
+                    let fenv = undefined
```

Comment:
Minor nit: can we turn these `undefined` calls into `error` calls with a more descriptive message (e.g., "Formula parsing environment not supported in the test environment") in case something changes to trigger this later?